### PR TITLE
migrate(`forge-bind`): to alloy

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Final part of alloy migration

## Solution

In order to do this, we have to make use of parts of code that make up the `sol!` macro to create a `SolMacroGen` type that is similar to the `Abigen` type in ethers. 

We're reading the JSON abi files (the artifacts) and retrieving the `ABI` from it then converting that to a Solidity `TokenStream` and ultimately convert that to a `SolInput` that should be passed to the expander that is used by the `sol!` macro here: https://github.com/alloy-rs/core/blob/85aecdf54c56da20f274a21f372d3d37b8c118ca/crates/sol-macro/src/lib.rs#L281.

However the issue is that the `expand` module is private in the `sol-macro` crate and cannot be made public since it's not a macro. 

@DaniPopes @mattsse I can move the `expand` module to a new crate `sol-macro-expander`, make it public and use it in foundry??

TODO
- [x] Artifacts and files filter
- [x] Make `expand` public by moving to a new crate.
- [x] Write to crate and module

### CODE REVIEW

**Comment by yash-atreya** [comment]
Proposing to make the utils method `tokens_for_sol` https://github.com/alloy-rs/core/blob/85aecdf54c56da20f274a21f372d3d37b8c118ca/crates/sol-macro-input/src/json.rs#L74 also public in `alloy-rs/core`, we can then remove the `syn`, `proc_macro2` deps.

**Comment by yash-atreya** [comment]
TODO:

- [ ] `Filter` for selecting and excluding ABIs in alloy
- [ ] Adding `derive` and `sol` attributes

**Review by mattsse — COMMENTED** [review]
@DaniPopes is best equipped to give directions here, but I think this sounds reasonable

File: crates/forge/bin/cmd/bind.rs [hunk 1]
```diff
@@ -84,7 +85,7 @@ pub struct BindArgs {
     skip_extra_derives: bool,
 
     /// Generate bindings for the `alloy` library, instead of `ethers`.
-    #[arg(skip)]
+    #[arg(long)]
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind.rs:88]:
```suggestion
    #[arg(long, conflicts_with = "ethers")]
```

File: crates/forge/src/sol_macro_gen.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    pub expansion: Option<TokenStream>,
+}
+
+impl SolMacroGen {
+    pub fn new(path: PathBuf, name: String) -> Self {
+        Self { path, name, expansion: None }
+    }
+
+    pub fn get_json_abi(&self) -> (JsonAbi, Option<String>) {
+        let json = std::fs::read(&self.path).expect("Failed to read JSON file");
```
Comment (DaniPopes) [inline crates/forge/src/sol_macro_gen.rs:20]: should be errors instead of panics

File: crates/forge/src/sol_macro_gen.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    ) {
+        self.generate_bindings();
+
+        let src = bindings_path.join("src");
+
+        let _ = fs::create_dir_all(&src);
+
+        // Write Cargo.toml
+        let cargo_toml_path = bindings_path.join("Cargo.toml");
+        let toml_contents = format!("[package]\nname = \"{}\"\nversion = \"{}\"\nedition = \"2021\"\n\n[dependencies]\nalloy-sol-types = \"0.7.4\"\nalloy-contract = {{ git = \"https://github.com/alloy-rs/alloy\", rev = \"64feb9b\" }}", name, version);
```
Comment (DaniPopes) [inline crates/forge/src/sol_macro_gen.rs:157]: pls make these strings readable with `r#""#`

File: crates/forge/src/sol_macro_gen.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    pub instances: Vec<SolMacroGen>,
+}
+
+impl MultiSolMacroGen {
+    pub fn new(artifacts_path: &Path) -> Self {
+        let artifacts_path = artifacts_path.to_path_buf();
+        let mut instances = Vec::new();
+
+        let abi_files = json_files(&artifacts_path)
+            .filter_map(|path| {
```
Comment (DaniPopes) [inline crates/forge/src/sol_macro_gen.rs:56]: This is already done in `BindArgs::get_json_files`, pls reuse code

**Comment by yash-atreya** [comment]
@DaniPopes please drop your final review whenever possible

File: crates/forge/bin/cmd/bind.rs [hunk 5]
```diff
@@ -84,7 +85,7 @@ pub struct BindArgs {
     skip_extra_derives: bool,
 
     /// Generate bindings for the `alloy` library, instead of `ethers`.
-    #[arg(skip)]
+    #[arg(long, conflicts_with = "ethers")]
     alloy: bool,
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind.rs:89]: replace the eprintln with the commented one in line 108

File: crates/forge/bin/cmd/bind.rs [hunk 6]
```diff
@@ -158,9 +159,29 @@ impl BindArgs {
             .into()
     }
 
+    fn get_alloy_filter(&self) -> Filter {
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind.rs:162]: is this necessary? The filtering and file finding behavior should be the same

File: crates/forge/bin/cmd/bind.rs [hunk 7]
```diff
@@ -239,13 +281,29 @@ impl BindArgs {
         Ok(())
     }
 
+    fn check_alloy(&self, artifacts: &Path, bindings_root: &Path) -> Result<()> {
+        let mut bindings = self.get_solmacrogen(artifacts)?;
+        bindings.generate_bindings()?;
+        println!("Checking bindings for {} contracts", bindings.instances.len());
+        let _ = bindings.check_consistency(
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind.rs:288]: the error is ignored

File: crates/forge/bin/cmd/bind.rs [hunk 8]
```diff
@@ -267,4 +325,60 @@ impl BindArgs {
...
+            trace!(single_file = self.single_file, "generating crate");
+            let _ = solmacrogen.write_to_crate(
+                &self.crate_name,
+                &self.crate_version,
+                bindings_root,
+                self.single_file,
+            );
+        } else {
+            trace!(single_file = self.single_file, "generating module");
+            let _ = solmacrogen.write_to_module(bindings_root, self.single_file);
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/bind.rs:343]: the error is ignored

File: crates/forge/Cargo.toml [hunk 9]
```diff
@@ -60,9 +60,14 @@ alloy-provider = { workspace = true, features = ["reqwest", "ws", "ipc"] }
...
 alloy-transport.workspace = true
 alloy-signer.workspace = true
+alloy-sol-macro-input.workspace = true
+alloy-sol-macro-expander = { workspace = true, features = ["json"] }
 alloy-consensus.workspace = true
 alloy-chains.workspace = true
 
+proc-macro2 = "1.0.82"
+quote = "1.0"
+syn = "2.0"
```
Comment (DaniPopes) [inline crates/forge/Cargo.toml:70]: these should be in the workspace

File: crates/forge/src/lib.rs [hunk 10]
```diff
@@ -23,6 +23,8 @@ pub use runner::ContractRunner;
 
 pub mod result;
 
+pub mod sol_macro_gen;
```
Comment (DaniPopes) [inline crates/forge/src/lib.rs:26]: pls extract to a separate crate

File: crates/forge/src/sol_macro_gen.rs [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+alloy-contract = {{ git = "https://github.com/alloy-rs/alloy", rev = "64feb9b" }}"#,
+            name, version
+        );
+
+        fs::write(cargo_toml_path, toml_contents)
+            .map_err(|e| eyre::eyre!("Failed to write Cargo.toml: {e}"))?;
+
+        // Write src
+        let mut lib_contents = if single_file {
+            String::from("#![allow(unused_imports, clippy::all)]\n\n//! This module contains the sol! generated bindings for solidity contracts.\n//! This is autogenerated code.\n//! Do not manually edit these files.\n//! These files may be overwritten by the codegen system at any time.\n")
```
Comment (DaniPopes) [inline crates/forge/src/sol_macro_gen.rs:141]: pls make all these strings multiline like above

File: crates/forge/src/sol_macro_gen.rs [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+        let src = bindings_path.join("src");
+
+        let _ = fs::create_dir_all(&src);
+
+        // Write Cargo.toml
+        let cargo_toml_path = bindings_path.join("Cargo.toml");
+        let toml_contents = format!(
+            r#"
+[package]
```
Comment (DaniPopes) [inline crates/forge/src/sol_macro_gen.rs:125]:
```suggestion
            r#"[package]
```

File: crates/forge/src/sol_macro_gen.rs [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+name = "{}"
+version = "{}"
+edition = "2021"
+
+[dependencies]
+alloy-sol-types = "0.7.4"
+alloy-contract = {{ git = "https://github.com/alloy-rs/alloy", rev = "64feb9b" }}"#,
```
Comment (DaniPopes) [inline crates/forge/src/sol_macro_gen.rs:132]: there should be an option to pass the alloy version

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/forge/bin/cmd/bind.rs (hunk 6 above)
Comment (yash-atreya) [reply crates/forge/bin/cmd/bind.rs:162]: Yeah, so `get_filter` returns `ContractFilter` which is specific to `ethers`.  These methods just return the regex that is then used in a unified setting in `get_json_files` to actually get the files.

**Comment by yash-atreya** [comment]
@DaniPopes ptal

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+//!
+//! This type encapsulates the logic for expansion of a Rust TokenStream from Solidity tokens. It
+//! uses the `expand` method from `alloy_sol_macro_expander` underneath.
+//!
+//! It holds info such as `path` to the ABI file, `name` of the file and the rust binding being
+//! generated, and lastly the `expansion` itself, i.e the Rust binding for the provided ABI.
+//!
+//! It contains methods to read the json abi, generate rust bindings from the abi and ultimately
+//! write the bindings to a crate or modules.
+use alloy_json_abi::JsonAbi;
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:11]:
```suggestion
//! write the bindings to a crate or modules.

use alloy_json_abi::JsonAbi;
```

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+                    name
+                ) + &instance.expansion.as_ref().unwrap().to_string() +
+                    "\n\n";
+                mod_contents += &contents;
+            }
+        }
+
+        let mod_path = bindings_path.join("mod.rs");
+        fs::write(mod_path, mod_contents)
+            .map_err(|e| eyre::eyre!("Failed to write mod.rs: {e}"))?;
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:223]: code should be formatted before writing to file

**Review by DaniPopes — CHANGES_REQUESTED** [review]
can you run this on a few popular users (just search "forge bind" on github) and check if the output makes sense

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+            let sol_str = json_abi.to_sol(&instance.name, None);
+
+            let ident_name: Ident = Ident::new(&instance.name, Span::call_site());
+
+            let tokens = tokens_for_sol(&ident_name, &sol_str)
+                .map_err(|e| eyre::eyre!("Failed to get sol tokens: {e}"))?;
+
+            let tokens = quote::quote! {
+                #[derive(Debug)]
+                #[sol(rpc)]
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:87]: This is missing bytecode attributes

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+        );
+        for instance in &self.instances {
+            let name = instance.name.to_lowercase();
+            if !single_file {
+                mod_contents += &format!(
+                    r#"pub mod {};
+                "#,
+                    instance.name.to_lowercase()
+                );
+                let contents = "//! This module was autogenerated by the alloy sol!.\n//! More information can be found here <https://docs.rs/alloy-sol-macro/latest/alloy_sol_macro/macro.sol.html>.\n".to_string() + &instance.expansion.as_ref().unwrap().to_string();
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:213]: use format! and split to multiple lines like above

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+        let sol_types_check = &format!("alloy-sol-types = \"{}\"", alloy_version);
+        let alloy_contract_check =
+            "alloy-contract = {{ git = \"https://github.com/alloy-rs/alloy\", rev = \"64feb9b\" }}";
+        let toml_consistent = cargo_toml_contents.contains(name_check) &&
+            cargo_toml_contents.contains(version_check) &&
+            cargo_toml_contents.contains(sol_types_check) &&
+            cargo_toml_contents.contains(alloy_contract_check);
+        eyre::ensure!(
+                toml_consistent,
+                format!("The contents of Cargo.toml do not match the expected output of the newest `ethers::Abigen` version.\
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:347]: this error message can be 2 lines

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Comment by yash-atreya** [comment]
@DaniPopes merge?

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+        let cargo_toml_path = bindings_path.join("Cargo.toml");
+        let toml_contents = format!(
+            r#"[package]
+name = "{}"
+version = "{}"
+edition = "2021"
+
+[dependencies]
+alloy-sol-types = "{}"
+alloy-contract = {{ git = "https://github.com/alloy-rs/alloy", rev = "64feb9b" }}"#,
```
Comment (mattsse) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:144]: this is bad, should not hardcode rev
```suggestion
alloy-contract = {{ git = "https://github.com/alloy-rs/alloy" }}"#,
```

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+                //! This module contains the sol! generated bindings for solidity contracts.
+                //! This is autogenerated code.
+                //! Do not manually edit these files.
+                //! These files may be overwritten by the codegen system at any time."#,
+            )
+        } else {
+            String::from(
+                r#"#![allow(unused_imports)]
+            "#,
+            )
```
Comment (mattsse) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:164]: .to_string

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+
+        eyre::ensure!(cargo_toml_path.is_file(), "Cargo.toml must exist");
+        let cargo_toml_contents = fs::read_to_string(cargo_toml_path)
+            .map_err(|e| eyre::eyre!("Failed to read Cargo.toml: {e}"))?;
+
+        let name_check = &format!("name = \"{}\"", name);
+        let version_check = &format!("version = \"{}\"", version);
+        let sol_types_check = &format!("alloy-sol-types = \"{}\"", alloy_version);
+        let alloy_contract_check =
+            "alloy-contract = {{ git = \"https://github.com/alloy-rs/alloy\", rev = \"64feb9b\" }}";
```
Comment (mattsse) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:354]: no rev please

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+    pub fn get_json_abi(&self) -> Result<(JsonAbi, Option<String>)> {
+        let json = std::fs::read(&self.path)?;
+
+        // Need to do this to get the abi in the next step.
+        let json: Value = serde_json::from_slice(&json)?;
+
+        let abi_val = json.get("abi").ok_or_eyre("No ABI found in JSON file")?;
+        let json_abi = serde_json::from_str(&abi_val.clone().to_string())?;
+
+        let bytecode = json.get("bytecode").and_then(|b| b.get("object")).map(|o| o.to_string());
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:44]: `JsonAbi` can be deserialized directly. Please see the logic in sol-macro-input

in fact I would prefer if you just passed the path through `quote!` to syn and parse a SolInput directly

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 23]
```diff
@@ -0,0 +1,367 @@
...
+                let contents = prettyplease::unparse(&file);
+
+                fs::write(path, contents).map_err(|e| eyre::eyre!("Failed to write file: {e}"))?;
+                lib_contents += &format!(
+                    r#"pub mod {};
+                    "#,
+                    name
+                );
+            } else {
+                lib_contents += &contents;
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:183]: use `write!(lib, "{}")`

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 24]
```diff
@@ -0,0 +1,367 @@
...
+alloy-sol-types = "{}"
+alloy-contract = {{ git = "https://github.com/alloy-rs/alloy", rev = "64feb9b" }}"#,
+            name, version, alloy_version
+        );
+
+        fs::write(cargo_toml_path, toml_contents)
+            .map_err(|e| eyre::eyre!("Failed to write Cargo.toml: {e}"))?;
+
+        // Write src
+        let mut lib_contents = if single_file {
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:152]: i'd prefer `let mut = String::new();` and writing after

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 25]
```diff
@@ -0,0 +1,367 @@
...
+            } else {
+                let contents = format!(
+                    r#"pub use {}::*;
+                    //! This module was autogenerated by the alloy sol!.
+                    //! More information can be found here <https://docs.rs/alloy-sol-macro/latest/alloy_sol_macro/macro.sol.html>.
+                    "#,
+                    name
+                ) + &instance.expansion.as_ref().unwrap().to_string() +
+                    "\n\n";
+                mod_contents += &contents;
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:243]: same as above

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 26]
```diff
@@ -0,0 +1,367 @@
...
+            }
+        }
+
+        let mod_path = bindings_path.join("mod.rs");
+        let mod_file = syn::parse_file(&mod_contents)?;
+        let mod_contents = prettyplease::unparse(&mod_file);
+
+        fs::write(mod_path, mod_contents)
+            .map_err(|e| eyre::eyre!("Failed to write mod.rs: {e}"))?;
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:252]: this can be `wrap_err("Failed")`

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 27]
```diff
@@ -0,0 +1,367 @@
...
+
+                self.check_file_contents(&path, &tokens)?;
+
+                if !is_mod {
+                    super_contents += &format!(
+                        r#"pub mod {};
+                    "#,
+                        name
+                    );
+                }
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:309]: same as above

File: crates/sol-macro-gen/src/sol_macro_gen.rs [hunk 28]
```diff
@@ -0,0 +1,367 @@
...
+        let alloy_contract_check =
+            "alloy-contract = {{ git = \"https://github.com/alloy-rs/alloy\", rev = \"64feb9b\" }}";
+        let toml_consistent = cargo_toml_contents.contains(name_check) &&
+            cargo_toml_contents.contains(version_check) &&
+            cargo_toml_contents.contains(sol_types_check) &&
+            cargo_toml_contents.contains(alloy_contract_check);
+        eyre::ensure!(
+                toml_consistent,
+                format!("The contents of Cargo.toml do not match the expected output of the newest `ethers::Abigen` version.\n
+This indicates that the existing bindings are outdated and need to be generated again.")
```
Comment (DaniPopes) [inline crates/sol-macro-gen/src/sol_macro_gen.rs:362]: rustfmt bugs out on long literals, pls remove `format!` and put it on multiple lines; also it's mentioning ethers abigen

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Review by DaniPopes — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
nice work!

**Pull request merged**